File: Library/src/main/java/org/endeavourhealth/library/model/imq/Return.java
package org.endeavourhealth.library.model.imq;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import com.fasterxml.jackson.annotation.JsonSetter;
import org.endeavourhealth.library.model.tripletree.TTIriRef;
import org.endeavourhealth.library.vocabulary.VocabEnum;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

@JsonPropertyOrder({"node", "variable", "iri", "name", "function", "as"})
@JsonInclude(JsonInclude.Include.NON_DEFAULT)
public class Return implements Returnable{
  private String iri;
  private String nodeRef;
  private String name;
  private String propertyRef;
  private String pathRef;
  private boolean inverse;
  private FunctionClause function;
  private TTIriRef units;
  private String as;
  private List<Return> returx;
  private TTIriRef dataType;
  @JsonProperty("case")
  private Case case_;
  private String description;
  private String value;

  public String getValue() {
    return value;
  }
  public Return setValue(String value){
    this.value = value;
    return this;
  }

  public String getPathRef() {
    return pathRef;
  }

  public Return setPathRef(String pathRef) {
    this.pathRef = pathRef;
    return this;
  }

  public Return setName(String name) {
    this.name = name;
    return this;
  }

  @JsonProperty("case")
  public Case getCase() {
    return case_;
  }

  @JsonProperty("case")
  public Return setCase(Case case_) {
    this.case_ = case_;
    return this;
  }

  public Return case_(Consumer<Case> builder) {
    builder.accept(this.setCase(new Case()).getCase());
    return this;
  }

  public TTIriRef getDataType() {
    return dataType;
  }

  public Return setDataType(TTIriRef dataType) {
    this.dataType = dataType;
    return this;
  }

  public Return setNodeRef(String nodeRef) {
    this.nodeRef = nodeRef;
    return this;
  }


  @JsonProperty("return")
  public List<Return> getReturn() {
    return returx;
  }

  public Return setReturn(List<Return> returns) {
    this.returx = returns;
    return this;
  }

  public Return addReturn(Return return_) {
    if (this.returx == null)
      this.returx = new ArrayList<>();
    this.returx.add(return_);
    return this;
  }

  public Return return_(Consumer<Return> builder) {
    Return return_ = new Return();
    addReturn(return_);
    builder.accept(return_);
    return this;
  }

  @JsonSetter
  public Return setAs(String as) {
    this.as = as;
    return this;
  }

  public Return as(String as) {
    this.as = as;
    return this;
  }

  public Return setUnits(TTIriRef units) {
    this.units = units;
    return this;
  }

  public Return setFunction(FunctionClause function) {
    this.function = function;
    return this;
  }

  public Return function(Consumer<FunctionClause> builder) {
    builder.accept(setFunction(new FunctionClause()).getFunction());
    return this;
  }

  public Return setInverse(boolean inverse) {
    this.inverse = inverse;
    return this;
  }




  public Return setIri(String iri) {
    this.iri = iri;
    return this;
  }

  public Return setIri(VocabEnum iri) {
    this.iri = iri.toString();
    return this;
  }

  public Return setPropertyRef(String propertyRef) {
    this.propertyRef = propertyRef;
    return this;
  }


  public Return setDescription(String description) {
    this.description = description;
    return this;
  }

  public String getIri() {
    return iri;
  }

  public String getNodeRef() {
    return nodeRef;
  }

  public String getName() {
    return name;
  }

  public String getPropertyRef() {
    return propertyRef;
  }

  public boolean isInverse() {
    return inverse;
  }

  public FunctionClause getFunction() {
    return function;
  }

  public TTIriRef getUnits() {
    return this.units;
  }

  public String getAs() {
    return as;
  }

  public String getDescription() {
    return description;
  }
}

File: src/interfaces/Return.ts
import z from "zod";

import { CaseSchema } from "./Case";
import { FunctionClause, FunctionClauseSchema } from "./FunctionClause";
import { TTIriRef, TTIriRefSchema } from "./TTIriRef";

// export interface Return {
//   iri?: string;
//   name?: string;
//   function?: FunctionClause;
//   as?: string;
//   nodeRef?: string;
//   propertyRef?: string;
//   pathRef?: string;
//   inverse?: boolean;
//   units?: TTIriRef;
//   dataType?: TTIriRef;
//   semanticMap?: TTIriRef;
//   description?: string;
//   value?: string;
//   case?: Case;
//   return?: Return[];
// }

export const ReturnSchema = z.strictObject({
  iri: z.string().optional(),
  name: z.string().optional(),
  function: FunctionClauseSchema.optional(),
  as: z.string().optional(),
  nodeRef: z.string().optional(),
  propertyRef: z.string().optional(),
  pathRef: z.string().optional(),
  inverse: z.boolean().default(false),
  units: TTIriRefSchema.optional(),
  dataType: TTIriRefSchema.optional(),
  semanticMap: TTIriRefSchema.optional(),
  description: z.string().optional(),
  value: z.string().optional(),
  case: CaseSchema.optional(),
  get return(): z.ZodPrefault<z.ZodArray<typeof ReturnSchema>> {
    return z.array(ReturnSchema).prefault([]);
  }
});

export type Return = z.output<typeof ReturnSchema>;
